util/apiratelimiter/apiratelimiter.go

```diff
@@ -0,0 +1,72 @@
+package apiratelimiter
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+type ApiRateLimiter interface {
+	GetVisitor(ip string) *rate.Limiter
+}
+
+type visitor struct {
+	limiter  *rate.Limiter
+	lastSeen time.Time
+}
+
+type apiRateLimiter struct {
+	limit    int
+	burst    int
+	visitors map[string]*visitor
+	mu       *sync.Mutex
+}
+
+func NewApiRateLimiter(limit int, burst int) *apiRateLimiter {
+	// Create a map to hold the rate limiters for each visitor and a mutex.
+	var visitors = make(map[string]*visitor)
+	var mu sync.Mutex
+	return &apiRateLimiter{
+		limit:    limit,
+		burst:    burst,
+		visitors: visitors,
+		mu:       &mu,
+	}
+}
+
+// Retrieve and return the rate limiter for the current visitor if it
+// already exists. Otherwise create a new rate limiter and add it to
+// the visitors map, using the IP address as the key.
+func (r *apiRateLimiter) GetVisitor(ip string) *rate.Limiter {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	v, exists := r.visitors[ip]
+	if !exists {
```

Comment:
Could consider using an RWMutex to prevent all of the readers on line 45 from locking each other out...or maybe these calls are infrequent enough that it doesn't matter?

Reply:
good point, let me change that

Reply:
great!

Reply:
@alexec Would the sync.Map cause readers to lock each other out, though?

Reply:
Never mind. I see the documentation for sync.Map says "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."